### PR TITLE
fix(content): dispatch MonetizationEvent instead of CustomEvent

### PR DESCRIPTION
**Pull request by sidvishnoi**
## Context

Closes https://github.com/interledger/web-monetization-extension/issues/9

## Changes proposed in this pull request

1. Dispatch `MonetizationEvent` instead of `CustomEvent` with detail
2. Keep supporting `event.detail` access for backward compat, but add a warning on accessing for first time.
3. Rename internal event (for communication between content scripts) from `monetization-v2` to `__wm-ext-monetization`, along with `onmonetization-attribute-changed`.
4. Rename contentScript to polyfill
5. Turn contentScript code from  a string to regular TS file, so we can typecheck it like rest of code (~presently typechecking disabled on that file as out of scope of this PR~ added typechecking too)
6. Now, MonetizationEvent interface is available on global.

A sample code:
```html
<!doctype html>
<html lang="en">
  <head onmonetization="console.warn('head', event)">
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Web Monetization test</title>
    <link
      rel="monetization"
      href="https://ilp.rafiki.money/sidvishnoi-usd-2"
      onmonetization="console.log('inline', event)"
    />
  </head>
  <body >
    <h1>Web Monetization test</h1>
    <script>
      const link = document.querySelector('link[rel="monetization"]')
      // link.addEventListener('load', console.log)
      link.addEventListener('monetization', (event) => {
        console.log('1: event', event)
        console.log('2: event', event instanceof window.MonetizationEvent)
        console.log('3: event.detail', event.detail)
      })
      window.addEventListener('monetization', (ev) =>
        console.log('window:', ev)
      )
    </script>
  </body>
</html>
```

### CODE REVIEW

**Comment by raducristianpopa** [comment]
<h2>Extension builds preview</h2>

<table role="table">
<thead>
  <tr>
    <th>Name</th>
    <th>Link</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Latest commit</td>
    <td align="center">3b1224a7388d5ae56dcaf3d990819e8a2f2734ba</td>
  </tr>
  <tr>
    <td align="center">Latest job logs</td>
    <td align="center"><a href="https://github.com/interledger/web-monetization-extension/actions/runs/9546766229">Run #9546766229</a></td>
  </tr>
  <tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/24983300393/artifacts/1608182393">Download</a></td></tr><tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/24983300393/artifacts/1608182396">Download</a></td></tr>
</tbody>
</table>

**Comment by sidvishnoi** [comment]
Question:
https://webmonetization.org/docs/references/monetizationevent/ says we emit `amountSent` with value and currency. There's no assetScale there.
And how's `amountSent` different from `receiveAmount`?

**Comment by raducristianpopa** [comment]
> Question: https://webmonetization.org/docs/references/monetizationevent/ says we emit `amountSent` with value and currency. There's no assetScale there. And how's `amountSent` different from `receiveAmount`?

The current implementation is not following the spec. Based on the spec (https://webmonetization.org/specification/#amountsent-attribute), the currency should only be an [ISO4217](https://www.iso.org/iso-4217-currency-codes.html) string. I forgot about this part when sending the details about a monetization event from the background to the content script. We should convert the amount value to a user readable format:

```ts
// We should be careful about floating point precision as well
const amountSent = receiveAmount.value * 10 ** -receiveAmount.assetScale
```

The monetization event should only contain the following properties:
- `amountSent` - based on the [PaymentCurrencyAcomunt](https://www.w3.org/TR/payment-request/#dom-paymentcurrencyamount)
- `paymentPointer`
- `incomingPayment`

`receiveAmount` should not be present.

**Comment by sidvishnoi** [comment]
> the currency should only be an [ISO4217](https://www.iso.org/iso-4217-currency-codes.html) string

I think then the concern was crypto currencies not having those codes? Maybe something to figure out later?

**Comment by raducristianpopa** [comment]
> > the currency should only be an [ISO4217](https://www.iso.org/iso-4217-currency-codes.html) string
> 
> I think then the concern was crypto currencies not having those codes? Maybe something to figure out later?

This shouldn't really be an issue at the moment. Most ASE that enable WM for their user will probably run Rafiki, which has a constraint on the asset code as well - ISO4217: https://rafiki.dev/concepts/asset/.

File: src/content/services/contentScript.ts [hunk 1]
```diff
@@ -58,4 +59,14 @@ export class ContentScript {
       }
     )
   }
+
+  // TODO: When Firefox supports `world: MAIN`, inject this directly via
```
Comment (raducristianpopa) [inline src/content/services/contentScript.ts:63]: It looks like we might get this in the next release - Firefox 128? :eyes: 
https://bugzilla.mozilla.org/show_bug.cgi?id=1736575

What's going to be the issue if we add it with `world: 'MAIN'` right now?

File: src/content/polyfill.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      if (!eventDetailDeprecationEmitted) {
+        const msg = `MonetizationEvent.detail is deprecated. Access attributes directly instead.`
+        // eslint-disable-next-line no-console
+        console.warn(msg)
+        eventDetailDeprecationEmitted = true
+      }
+      const { receiveAmount, incomingPayment, paymentPointer } = this
+      return { receiveAmount, incomingPayment, paymentPointer }
+    }
+  }
```
Comment (raducristianpopa) [inline src/content/polyfill.ts]: Are there any tools that are using `event.details` to retrieve the information?

File: src/content/services/contentScript.ts (hunk 1 above)
Comment (sidvishnoi) [reply src/content/services/contentScript.ts:63]: Yes, I'm following that issue.

If we add `world: MAIN` now: install time warnings and `window.Monetization` not available.

Comment (sidvishnoi) [reply src/content/services/contentScript.ts:63]: (But we plan to support some older versions I think, so, might have to wait a year). I'll add the bugzilla link in a comment.

File: src/content/polyfill.ts (hunk 2 above)
Comment (sidvishnoi) [reply src/content/polyfill.ts]: I'm not aware, but to be on safer side.

File: src/content/services/monetizationTagManager.ts [hunk 3]
```diff
@@ -69,12 +70,18 @@ export class MonetizationTagManager extends EventEmitter {
...
-      const customEvent = new CustomEvent('monetization', {
-        bubbles: true,
-        detail: mozClone(details, this.document)
-      })
-
-      tag.dispatchEvent(customEvent)
+      const { incomingPayment, paymentPointer, receiveAmount } = details
+      const amountSent: PaymentCurrencyAmount = {
+        currency: receiveAmount.assetCode,
+        value: transformBalance(receiveAmount.value, receiveAmount.assetScale)
```
Comment (raducristianpopa) [inline src/content/services/monetizationTagManager.ts:76]: Not really important, but I would prefer sending the correct details directly from background. Thoughts?

File: src/shared/messages.ts [hunk 4]
```diff
@@ -131,7 +131,11 @@ export enum BackgroundToContentAction {
 
 export interface MonetizationEventPayload {
   requestId: string
-  details: any
+  details: {
```
Comment (raducristianpopa) [inline src/shared/messages.ts:134]: If transforming the data in background
```suggestion
  detail: {
```

File: src/content/services/monetizationTagManager.ts (hunk 3 above)
Comment (sidvishnoi) [reply src/content/services/monetizationTagManager.ts:76]: Oh yes, I didn't realise I was in content/

File: src/content/polyfill.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      monetizationTag.dispatchEvent(new MonetizationEvent(event.detail))
+    },
+    { capture: true }
+  )
+
+  window.addEventListener(
+    'onmonetization-attr-changed',
+    (event) => {
+      // eslint-disable-next-line no-console
+      console.log('onmonetization-attr-changed', event.detail)
```
Comment (raducristianpopa) [inline src/content/polyfill.ts:110]: I feel like we should not log this. If we want to keep it, let's make it a debug log.

Comment (sidvishnoi) [reply src/content/polyfill.ts:110]: Why is this event even fired though? I just left it as is for now.

Comment (sidvishnoi) [reply src/content/polyfill.ts:110]: I'm yet to check how this event supposed to work. Will do within this PR then.

Comment (raducristianpopa) [reply src/content/polyfill.ts:110]: It is used to register the `onmonetization` event handler for an element. Every time the handler is changed, we need to re-register the function.

```html
<link rel="monetization" ... onmonetization="console.log(event)" />
```

- When the page is loaded, we register the `onmonetization` event handler;
- If the website is changing the handler, we need to override the previous one;

Comment (sidvishnoi) [reply src/content/polyfill.ts]: Removed log.
Fixed bug: make sure event is fired and caught by polyfill (timing issue)
Refactor: renamed the internal event to start with `__wm_ext`.

File: src/content/polyfill.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    },
+    { capture: true }
+  )
+
+  window.addEventListener(
+    '__wm_ext_onmonetization_attr_change',
+    (event: CustomEvent<{ attribute?: string }>) => {
+      const link = event.target
+      if (!(link instanceof HTMLLinkElement)) return
+      if (!link.isConnected) return
```
Comment (raducristianpopa) [inline src/content/polyfill.ts:112]: The `onmonetization` event handler can be present on any element:

> The onmonetization event handler is an[ event handler content attribute](https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-content-attributes) that can be applied to any [element](https://dom.spec.whatwg.org/#concept-element). The user agent uses it to notify that some link has been [monetized](https://webmonetization.org/specification/#dfn-monetized).

Comment (sidvishnoi) [reply src/content/polyfill.ts:112]: Oh. Fixing..

Comment (sidvishnoi) [reply src/content/polyfill.ts:112]: btw, monetization event is on Line 95 above.

**Review by raducristianpopa — APPROVED** [review]

**Pull request merged**